### PR TITLE
Binary templates: include custom entries in section range

### LINKED ISSUE #242: Nodes created with `entry` will break the range property of any of its containing nodes.
In #178 it was at least partially acknowledged that the <code><b>entry</b></code> command is used not only for its intended purpose at its implementation – the representation of information that have no clear relationship to any region in the file – but instead, increasingly used as a sort of swiss-army knife of a command after its incident property of being able to represent arbitrary string values began to be exploited, or rather, employed in want of some other command meant to be customizable. Common applications include something as simple as limited formatting/styling of the represented value, to front-ends for complete parsers of entire missing data formats. 

An example would be parsing packed-bit formats of multiple data in orthogonal planes, such as the typical ARGB2101010 pixel format in a modern macOS framebuffer. The four values (Alpha, Red, Green, Blue) are stored in four bytes, but no single byte contains a single value, and no single value is stored entirely within a byte. After calculaitng the four values packed, or interleaved, across four bytes, the four values could intuitively be represented in a single-node as an ARGB tuple.

Using the full syntax of the command, <code><b>entry</b> <i>label ?length ?offset? ?</i></code> the node would have a custom range property as well, so it would behave more similarly to other synthesized nodes, highlighting the corresponding byte ranges while selected in the binary template.

One shortcoming that quickly becomes apparent when working with sections, especially multiple levels of nested sections, is that any section including a node created with the <code><b>entry</b></code> command will not seem to have a range property; no bytes will be highlighted when the boundary "container" node is selected.

#231 is my attempt to eliminate this inconsistency.

### PR DESCRIPTION
**Pull request by kumowoon1025**
When using the `entry` command with a <code><i>len</i></code> argument to add a node with a specified range, extend the range of the parent node to include the child node. (So it is reflected in the selected/highlighted range in the file when the entry is selected) Proposed as a way to 
Fixes #242 .

### CODE REVIEW

**Comment by kumowoon1025** [comment]
|Current behavior|Proposed change|
|----------------|-----------------|
|Current behavior|Proposed change|

The first top level atom is parsed using native "read type" commands only. Highlighting nested nodes works as expected.

The second top level atom has a node created with the `entry` command, specifying the entire range of the atom as its range. Specifying the range allows it to be highlighted, but currently, it breaks the range property of any of its containing nodes in the hierarchy. This would fix #242

File: app/sources/HFTemplateController.m [hunk 1]
```diff
@@ -526,15 +526,21 @@ - (HFTemplateNode *)currentSection {
...
-        node.range = HFRangeMake(self.anchor + self.position, *length);
+    HFTemplateNode *currentNode = self.currentNode;
+    HFTemplateNode *newNode = [[HFTemplateNode alloc] initWithLabel:label value:value];
+    if (length) {
+        if (offset) {
+            newNode.range = HFRangeMake(self.anchor + *offset, *length);
+        } else {
+            newNode.range = HFRangeMake(self.anchor + self.position, *length);
+        }
+        unsigned long long newloc = MIN(currentNode.range.location, newNode.range.location);
```
Comment (kumowoon1025) [inline app/sources/HFTemplateController.m:537]: In addition to making a new HFRange for the node being added, an additional HFRange is made by calculating the union of the range of the new node and the range of its parent; then the union is assigned as the parent's range property, reflecting the fact that it "contains" whatever range the new node is associated with.

**Comment by wader** [comment]
I used this change while developing #251. Seem to work fine.

**Comment by wader** [comment]
Was reading the suggestions in https://github.com/ridiculousfish/HexFiend/issues/193 ... maybe the selection or some kind of marker should always try to be in "sync" between the hex, ascii and template tree?

**Comment by kainjow** [comment]
Thanks!

**Pull request merged**